### ISSUE TITLE
[N64] Banjo-Tooie soft-locked at title screen on July 7th nightly

### ISSUE THREAD

**Issue by retrobenny** [higan-emu/ares#633]
From today's nightly build cb4a79e , the game now starts with the camera at the origin-point instead of doing the title screen sequence.
This also happens in Dillonb's emulator between updates.

I got the nightly build to test DK64 due to changes days after V128 was released, but its on a black screen so it doesn't even boot now which is also what happened between updates on Dillonb's emulator as well.

**Comment by mystise** [comment]
Bisected to https://github.com/ares-emulator/ares/commit/57b4d682986405fbc3271c502ebd074767636216

Same as https://github.com/ares-emulator/ares/issues/605

~~Note that that commit is correct, moving the queue access after the dma always produces 0x7F as the read / write length (barring minor exceptions), but the cycle-per-byte count is too long for Tooie and DK64.~~

~~Given that the timing was initially changed in https://github.com/ares-emulator/ares/commit/329a8c958b0f0e742ea871727918e61dcbf44e64 to better account for proper DMA timings, however I suspect that 9 COP0 cycles is a bit too long in the first place.~~

~~The only other source I could find for PI DMA timings was the MAME source.~~

~~https://github.com/mamedev/mame/blob/ab6237da824a95898209ecbb568095981c4dbe69/src/mame/nintendo/n64_m.cpp#L1609~~

~~It looks like their measurements show 2.54 COP0 cycles = 5.08 CPU cycles = 10.16 Ares cycles per byte.~~

~~Matching the MAME value of 10 fixes both Banjo Tooie and DK64 without breaking Cruis'n Exotica.~~

~~Edit:
Note that the relevant CEN64 implementation has a hack comment next to it and completes the first tick after a DMA is requested.~~

~~https://github.com/n64dev/cen64/blob/d386930fa2d12d874c7631ab0786dfde00df1606/pi/controller.c#L37~~

~~During testing the maximum value that Banjo Tooie and DK64 would boot with was 23 ares cycles = 11.5 CPU cycles = 5.75 COP0 cycles, any higher and Banjo Tooie and DK64 both fail to load.~~

Edit:
See https://github.com/ares-emulator/ares/pull/931#issuecomment-1372326149 and https://github.com/ares-emulator/ares/pull/931#issuecomment-1372198894 for more accurate information.

**Comment by invertego** [comment]
Luke and rasky covered much of this in comments on mystise's PR (#931), but I just posted the root cause analysis I did last year which contains some additional details: https://github.com/ares-emulator/ares/issues/605#issuecomment-1374378855

It was conducted on DK64 but it should apply to Banjo Tooie as well.

@mystise: I'm sorry I didn't share this sooner; it would have saved you some duplicated effort. That you took the time to look into this is still very much appreciated!

**Comment by mystise** [comment]
After looking more at the Banjo Tooie (cpu) code, it never does a DMA of more than 0x20000 bytes at a time (it explicitly splits up large DMAs into chunks), so it might be a slightly different issue (but still caused by DMA timings)

I don't have the RSP code on hand so it might be doing something different to DK64 that causes the same race condition at a lower byte count, or a different procedure in the RSP might be racing

**Comment by mystise** [comment]
Ah! It turns out this is actually a race with the CIC 6105 anti-piracy check:

At some point during execution, the 6105 boot code on the RSP populates some offsets within memory with copies from its instruction memory.

(see https://github.com/mupen64plus/mupen64plus-rsp-hle/blob/master/src/cicx105.c for a high level replication of how the bytes get written)

The CIC thread reads those values, xors them, and then if the result is non-zero it corrupts certain objects on the heap.

The function that reads values from memory is at 0x8001de64, which reads from 0xa02fe1c0 and 0xa02fb1f4.

The expected values at those addresses are

```
*(uint_32_t*)0xa02fe1c0 = 0xad170014
*(uint_32_t*)0xa02fb1f4 = 0xad090010
```

However only the low byte is used (xored with 0x14 and 0x10 respectively, then ored together)

I still don't have the RSP code on hand, nor do I know how to disassemble IPL3, but this should hopefully point things in the right direction.

**Comment by invertego** [comment]
The code that is relevant to the race is the same as DK64. You can verify this with ares' built-in tracing feature, though getting instruction traces does require that you compile from source with the interpreter enabled.

Note that this is all initiated by the IPL3 which is the same for both games.

PI DMA:
```
PI I/O: PI_DRAM_ADDRESS <= 00000400
PI I/O: PI_PBUS_ADDRESS <= 10001000
PI I/O: PI_WRITE_LENGTH <= 000fffff
```

RSP loop:
```
RSP  1e4  lui     a1,$0020
RSP  1e8  bltz    a1{$00200000},$258
RSP  1ec  mfc0    v1,SP_SEMAPHORE{$00000001}
RSP  1f0  bne     v1{$00000001},0,$1e8
RSP  1f4  subi    a1,a1{$00200000},$1
[Omitted: 8388610]
RSP  258  break
```

If you'd like to chat more directly, feel free to join the ares discord: https://discord.com/invite/gz2quhk2kv

**Comment by mystise** [comment]
Ah I see, so the initial DMA of the first MB of rom into memory is racing within the IPL3 code itself, so if it times out it never runs the rest of the IPL3 boot code, thus failing one of Tooie's anti-piracy checks despite successfully loading the initial map (which is itself a check because the map files are xored with the CIC challenge response codes)

Neat

Thanks for the info!

**Comment by Mr-Wiseguy** [comment]
Are there any updates on this issue? The latest release of Ares (132) still gets stuck at the intro cutscene.

**Comment by rasky** [comment]
Fixed by #1147
